src/coreComponents/physicsSolvers/fluidFlow/ThermalSinglePhaseFVMKernels.hpp

```diff
@@ -476,6 +476,353 @@ class FaceBasedAssemblyKernelFactory
+
+      // Compute the (upwinded) energy flux
+
+      real64 const flux = mobility_up * f;
+      real64 const enthalpy = m_enthalpy[er][esr][ei][0];
+      stack.energyFlux += flux * enthalpy;
+
+      // Compute the derivatives of the (upwinded) energy flux wrt pressure and temperature
+
+      if( f >= 0 ) // the element is upstream
+      {
+        real64 const dFlux_dP = mobility_up * dF_dP + dMobility_dP_up * f;
+        real64 const dFlux_dT = mobility_up * dF_dT + m_dMob_dTemp[er][esr][ei] * f;
+
+        stack.dEnergyFlux_dP += dFlux_dP * enthalpy + flux * m_dEnthalpy_dPres[er][esr][ei][0];
+        stack.dEnergyFlux_dT += dFlux_dT * enthalpy + flux * m_dEnthalpy_dTemp[er][esr][ei][0];
+      }
+      else
+      {
+        real64 const dFlux_dP = mobility_up * dF_dP;
+        real64 const dFlux_dT = mobility_up * dF_dT;
+
+        stack.dEnergyFlux_dP += dFlux_dP * enthalpy;
+        stack.dEnergyFlux_dT += dFlux_dT * enthalpy;
+      }
+
+      // Contribution of energy conduction through the solid phase
+      real64 thermalTrans = 0.0;
+      real64 dThermalTrans_dPerm[3]{}; // not used
+      m_stencilWrapper.computeWeights( iconn,
+                                       m_thermalConductivity,
+                                       thermalTrans,
+                                       dThermalTrans_dPerm );
+
+      stack.energyFlux += thermalTrans * ( m_temp[er][esr][ei] - m_faceTemp[kf] );
+      stack.dEnergyFlux_dT += thermalTrans;
+
+      // Add energyFlux and its derivatives to localFlux and localFluxJacobian
+      integer const localRowIndexEnergy = numEqn - 1;
+      stack.localFlux[localRowIndexEnergy] =  m_dt * stack.energyFlux;
+
+      stack.localFluxJacobian[localRowIndexEnergy][0] =  m_dt * stack.dEnergyFlux_dP;
+      stack.localFluxJacobian[localRowIndexEnergy][numDof-1] =  m_dt * stack.dEnergyFlux_dT;
+    } );
+
+  }
+
+  /**
+   * @brief Performs the complete phase for the kernel.
+   * @param[in] iconn the connection index
+   * @param[inout] stack the stack variables
+   */
+  GEOSX_HOST_DEVICE
+  void complete( localIndex const iconn,
+                 StackVariables & stack ) const
+  {
+    Base::complete( iconn, stack, [&] ( localIndex const localRow )
+    {
+      RAJA::atomicAdd( parallelDeviceAtomic{}, &AbstractBase::m_localRhs[localRow + numEqn - 1], stack.localFlux[numEqn-1] );
+
+      AbstractBase::m_localMatrix.addToRowBinarySearchUnsorted< parallelDeviceAtomic >
+        ( localRow + numEqn - 1,
+        stack.dofColIndices,
+        stack.localFluxJacobian[numEqn-1],
+        numDof );
+    } );
+  }
+
+protected:
+
+  /// Views on temperature
+  ElementViewConst< arrayView1d< real64 const > > const m_temp;
+
+  /// Views on face temperature
+  arrayView1d< real64 const > const m_faceTemp;
+
+  /// Views on derivatives of fluid mobilities
+  ElementViewConst< arrayView1d< real64 const > > const m_dMob_dPres;
```

Comment:
I don't think you need this member variable either, but I may be wrong
```suggestion
```

Reply:
done